usr.bin/utoppya/utoppya.1
.\" $NetBSD$
.\"
.\" Copyright (c) 2006,2015 The NetBSD Foundation, Inc.
.\" All rights reserved.
.\"
.\" This code is derived from software contributed to The NetBSD Foundation
.\" by Steve C. Woodford.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
.\" ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
.\" TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
.\" PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
.\" BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
.\" CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd April 3, 2006
.Dt UTOPPYA 1
.Os
.Sh NAME
.Nm utoppya
.Nd Topfield TF5000PVR DVB recorder file manipulation program
.Sh SYNOPSIS
.Nm
.Op Fl f Ar device
.Ar command ...
.Sh DESCRIPTION
.Nm
is the userland interface to the
Topfield TF5000PVR range of DVB recorders (nicknamed `Toppy')
which uses the
.Xr utoppy 4
device driver.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl f
Communicate with
.Ar device
instead of the default
.Pa /dev/utoppy0 .
.El
.Pp
Using the services of the
.Xr utoppy 4
driver,
.Nm
can perform the following operations, specified by the
.Ar command
operand:
.Bl -tag -width Ds
.It Cm df
Display disk size and free space
.It Cm ls Op Ar directory
List the files in the specified
.Ar directory .
Default is the Toppy's root directory.
.It Cm mkdir Ao Ar directory Ac
Make the specified
.Ar directory .
You must supply the full pathname to
.Ar directory .
.It Cm rm Ao Ar pathname Ac
Delete the file or directory specified by
.Ar pathname .
.It Cm rename Ao Ar source Ac Ao Ar target Ac
Rename the file or directory specified by the
.Ar source
operand to the destination file or directory specified by the
.Ar target
operand.
.It Cm get Oo Fl prt Oc Ao Ar toppyfile Ac Oo Ar localfile Oc
Copy
.Ar toppyfile
from the Toppy to
.Ar localfile
on the local file system.
If
.Ar localfile
is omitted, the file will be copied into the current directory and will be
named using the last component of the
.Ar toppyfile
operand.
If
.Ar localfile
is
.Sq - ,
then
.Ar toppyfile
will be copied to the standard output.
.Pp
The following options are available for the
.Nm get
command:
.Bl -tag -width Ds
.It Fl p
Display a progress bar.
.It Fl r
This option is useful if you wish to resume a previously interrupted
.Nm get
command.
Instead of restarting from the beginning of the file, the transfer will
resume where it left off.
.It Fl t
Enable
.Ql Turbo
mode.
This instructs the Toppy to drop everything and concentrate on transferring
the file as quickly as possible.
You will be able to watch live TV, but all other functions, including
changing channel via the remote control, will be inoperative.
.El
.It Cm put Oo Fl prt Oc Ao Ar localfile Ac Ao Ar toppyfile Ac
Copy
.Ar localfile
on the local filesystem to
.Ar toppyfile
on the Toppy.
If
.Ar toppyfile
specifies a directory on the Toppy, the last component of the
.Ar localfile
operand will be appended to the
.Ar toppyfile
operand.
.Pp
The options described for the
.Nm get
command (above) also apply to the
.Nm put
command.
.El
.Sh FILES
.Pa /dev/utoppy0
The default Topfield TF5000PVR instance.
.Sh SEE ALSO
.Xr usb 4 ,
.Xr utoppy 4
.Sh HISTORY
The
.Nm
command first appeared in
.Nx 4.0
and was inspired by
.Ql ftpd-topfield
written by Steve Bennett
.Aq msteveb at ozemail.com.au
and
.Ql puppy
written by Peter Urbanec
.Aq toppy at urbanec.net .
.Sh AUTHORS
.An Steve C. Woodford Aq Mt scw@netbsd.org


doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision$>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------


Changes from NetBSD 7.0 to NetBSD 8.0:
	libc: Update to tzcode2014f. [christos 20140815]
	gmake:	Update to gmake-3.81 in external/gpl2/gmake and remove 
		gnu/dist/gmake (3.80) [christos 20140818]
	module(7): Create MODULAR_DEFAULT_AUTOLOAD config flag to control
		whether module autoloading is on by default.
		[jnemeth 20140824]
	kernel: Option SYMTAB_SPACE has been replaced with COPY_SYMTAB.
		The related changes to src/sys/conf require a rebuild of
		config(1). [joerg 20140824]
	zoneinfo: Import tzdata2014g. [apb 20140831]
	wm(4): Add internal SERDES support for 82575 and newer controllers.
		[msaitoh 20140903]
	or1k: Add new port.  [matt 20140903]
	awindma(4): Add support for Allwinner A10/A20 DMA controller.
		[jmcneill 20140906]
	awinac(4): Add support for Allwinner A10/A20 audio codec.
		[jmcneill 20140906]
	awinrtc(4): Add support for Allwinner A10/A20 real-time clock.
		[jmcneill 20140907]
	awinmmc(4): Use DMA for MMC transfers. [jmcneill 20140908]
	axp20x(4): Add AXP20x PMU driver. [jmcneill 20140909]
	rtsol(8): Removed in favour of dhcpcd. [roy 20140911]
	rtsold(8): Removed in favour of dhcpcd. [roy 20140911]
	dhcpcd(8): Import dhcpcd-6.4.5. [roy 20140918]
	arm: Add support for i.MX6 SoC. [ryo 20140925]
	gpt(8): Completed overhaul, including adding follwing subcommands:
		resize, set, unset, backup, restore, and resizedisk.
		[jnemeth 20140926]
	dhcpcd(8): Import dhcpcd-6.4.7. [roy 20140927]
	mpt(4): added bio(4) support. [jmcneill 20140927]
	gpt(8): added type subcommand to modify partition type.
		[jnemeth 20140927]
	zoneinfo: Import tzdata2014h. [apb 20141004]
	dhcpcd(8): Import dhcpcd-6.5.0. [roy 20141006]
	libc: Update to tzcode2014h. [christos 20141007]
	file(1): Upgraded to 5.20. [christos 20141010]
	arm: Add support for Allwinner A31 SoC. [jmcneill 20141010]
	awiniic(4): Add support for Allwinner A31 SoC. [jmcneill 20141012]
	awinp2wi(4): Add support for Allwinner A31 Push-Pull Two Wire
		Interface. [jmcneill 20141012]
	axp22x(4): Add AXP22x PMU driver. [jmcneill 20141012]
	awinrtc(4): Add support for Allwinner A31 real-time clock.
		[jmcneill 20141012]
	awindma(4): Add support for Allwinner A31 DMA controller.
		[jmcneill 20141013]
	awinac(4): Add support for Allwinner A31 audio codec.
		[jmcneill 20141013]
	motg(4): Add support for Allwinner A10/A20 SoC. [jmcneill 20141015]
	wpa: Import wpa_supplicant and hostapd 2.3. [christos 20141016]
	dhcpcd(8): Import dhcpcd-6.5.1. [roy 20141018]
	OpenSSH: Imported 6.7. [christos 20141018]
	resolvconf(8): Import openresolv-3.6.0 [roy 20141020]
	gcc: Support -fsanitize=address [christos 20141022]
	zoneinfo: Import tzdata2014i. [apb 20141023]
	libc: Update to tzcode2014i. [christos 20141023]
	openpam(3): update to 20140912 (ourouparia) [christos 20141024]
	pppd(8): updated to version 2.4.7. [christos 20141025]
	acpi(4): Updated ACPICA to 20140926. [christos 20141025]
	dhcpcd(8): Import dhcpcd-6.6.0 [roy 20141029]
	resolvconf(8): Import openresolv-3.6.1 [roy 20141029]
	rtsx(4): Add support for Realtek RTS5227/RTL8402/RTL8411/RTL8411B.
		[nonaka 20141029]
	flex(1): Import flex-2.5.37 [christos 20141029]
	iwn(4): Add support for Intel Centrino Wireless-N
		100/105/130/135/2200/2230. From OpenBSD. [nonaka 20141030]
	awge(4): Add support for gigabit ethernet found on various
		Allwinner SOCs. [martin 20141101]
	dhcpcd(8): Import dhcpcd-6.6.1. [roy 20141107]
	awinhdmi(4): Add support for Allwinner A20/A31 HDMI controller.
		[jmcneill 20141110]
	awintcon(4): Add support for Allwinner A20/A31 TV/LCD controller.
		[jmcneill 20141110]
	awindebe(4): Add support for Allwinner A20/A31 Display engine
		backend (DE-BE). [jmcneill 20141110]
	awinhdmiaudio(4): Add support for Allwinner A20/A31 HDMI audio
		controller. [jmcneill 20141111]
	dhcpcd(8): Import dhcpcd-6.6.2. [roy 20141114]
	awinir(4): Add support for Allwinner A20/A31 IR receiver.
		[jmcneill 20141115]
	zoneinfo: Import tzdata2014j. [apb 20141116]
	libc: Change arc4random(3) to use ChaCha20 and per-thread state.
		[riastradh 20141116]
	libc: Update to tzcode2014j. [christos 20141117]
	libpcap: Import 1.6.2. [christos 20141119]
	tcpdump(8): Import 4.6.2. [christos 20141119]
	dhcpcd(8): Import dhcpcd-6.6.4. [roy 20141126]
	dhcpcd(8): Import dhcpcd-6.6.5. [roy 20141209]
	bind: Import version 9.10.1-P1. [christos 20141209]
	dhcpcd(8): Import dhcpcd-6.6.6. [roy 20141217]
	powerpc: Switch to GCC 4.8. [matt 20141218]
	dhcpcd(8): Import dhcpcd-6.6.7. [roy 20141219]
	ntp: Import ntp 4.2.8. [christos 20141219]
	midirecord(1): Add program to record MIDI files.  [mrg 20141229]
	file: update to 5.22 security limits [christos 20150102]
	byacc: update to 20141128 for more bison compat [christos 20150103]
	mgx: driver for Southland Media Systems (now Quantum 3D) MGX SBus
		graphics devices [macallan 20150106]
	gcc: Import GCC 4.8.4.  [mrg 20150109]
	openssl: Import openssl 1.0.1k [spz 20150113]
	am-utils: Upgrade to 6.2; adds nfsv4, lustre (which we can't use)
		nfs3 rpc support for amfs [christos 20150117]

usr.bin/utoppya/utoppya.c
/*	$NetBSD$	*/

/*-
 * Copyright (c) 2006 The NetBSD Foundation, Inc.
 * All rights reserved.
 *
 * This code is derived from software contributed to The NetBSD Foundation
 * by Steve C. Woodford.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
 * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
 * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include <sys/types.h>
#include <sys/stat.h>
#include <sys/ioctl.h>

#include <err.h>
#include <errno.h>
#include <fcntl.h>
#include <libgen.h>
#include <sysexits.h>
#include <stdio.h>
#include <stdlib.h>
#include <strings.h>
#include <unistd.h>
#include <time.h>
#include <inttypes.h>

#include <dev/usb/utoppy.h>

#define	GLOBAL
#include "progressbar.h"

#define	_PATH_DEV_UTOPPY	"/dev/utoppy0"

/*
 * This looks weird for a reason. The toppy protocol allows for data to be
 * transferred in 65535-byte chunks only. Anything more than this has to be
 * split within the driver. The following value leaves enough space for the
 * packet header plus some alignmnent slop.
 */
#define	TOPPY_IO_SIZE	0xffec

static int toppy_fd;

static void cmd_df(int, char **);
static void cmd_ls(int, char **);
static void cmd_rm(int, char **);
static void cmd_mkdir(int, char **);
static void cmd_rename(int, char **);
static void cmd_get(int, char **);
static void cmd_put(int, char **);

static const struct toppy_command {
	const char *tc_cmd;
	void (*tc_handler)(int, char **);
} toppy_commands[] = {
	{"df",		cmd_df},
	{"ls",		cmd_ls},
	{"get",		cmd_get},
	{"mkdir",	cmd_mkdir},
	{"put",		cmd_put},
	{"rename",	cmd_rename},
	{"rm",		cmd_rm},
	{NULL,		NULL}
};

__dead static void
usage(void)
{

	fprintf(stderr, "usage: %s [-f <path>] <cmd> ...\n",
	    getprogname());

	exit(EX_USAGE);
}

int
main(int argc, char *argv[])
{
	const struct toppy_command *tc;
	const char *devpath;
	int ch;

	setprogname(argv[0]);
	devpath = _PATH_DEV_UTOPPY;

	while ((ch = getopt(argc, argv, "f:")) != -1) {
		switch (ch) {
		case 'f':
			devpath = optarg;
			break;

		default:
			usage();
		}
	}
	argc -= optind;
	argv += optind;

	if (argc == 0)
		usage();

	for (tc = toppy_commands; tc->tc_cmd != NULL; tc++)
		if (strcasecmp(argv[0], tc->tc_cmd) == 0)
			break;

	if (tc->tc_cmd == NULL)
		errx(EX_USAGE, "'%s' is not a valid command", argv[0]);

	if ((toppy_fd = open(devpath, O_RDWR)) < 0)
		err(EX_OSERR, "open(%s)", devpath);

	(*tc->tc_handler)(argc, argv);

	close(toppy_fd);

	return (0);
}

static int
find_toppy_dirent(const char *path, struct utoppy_dirent *udp)
{
	struct utoppy_dirent ud;
	char *d, *b, dir[FILENAME_MAX];
	ssize_t l;

	strncpy(dir, path, sizeof(dir));
	b = basename(dir);
	d = dirname(dir);
	if (strcmp(b, "/") == 0 || strcmp(b, ".") == 0 || strcmp(d, ".") == 0)
		errx(EX_USAGE, "'%s' is not a valid Toppy pathname", path);

	if (ioctl(toppy_fd, UTOPPYIOREADDIR, &d) < 0)
		err(EX_OSERR, "ioctl(UTOPPYIOREADDIR, %s)", d);

	if (udp == NULL)
		udp = &ud;

	while ((l = read(toppy_fd, udp, sizeof(*udp))) == sizeof(*udp)) {
		if (strcmp(b, udp->ud_path) == 0)
			break;
	}

	if (l < 0)
		err(EX_OSERR, "read(TOPPYDIR, %s)", d);

	if (l == 0)
		return (0);

	while (read(toppy_fd, &ud, sizeof(ud)) > 0)
		;

	return (1);
}

static void
cmd_df(int argc, char **argv)
{
	struct utoppy_stats us;

	if (ioctl(toppy_fd, UTOPPYIOSTATS, &us) < 0)
		err(EX_OSERR, "ioctl(UTOPPYIOSTATS)");

	printf("Hard Disk Size: %" PRId64 " MB\n", us.us_hdd_size / (1024 * 1024));
	printf("Hard Disk Free: %" PRId64 " MB\n", us.us_hdd_free / (1024 * 1024));
}

static void
cmd_ls(int argc, char **argv)
{
	struct utoppy_dirent ud;
	struct tm *tm;
	char *dir, *ext, dirbuf[2], ex, ft, tmbuf[32];
	ssize_t l;

	if (argc == 1) {
		dirbuf[0] = '/';
		dirbuf[1] = '\0';
		dir = dirbuf;
	} else
	if (argc == 2)
		dir = argv[1];
	else
		errx(EX_USAGE, "usage: ls [toppy-pathname]");

	if (ioctl(toppy_fd, UTOPPYIOREADDIR, &dir) < 0)
		err(EX_OSERR, "ioctl(UTOPPYIOREADDIR, %s)", dir);

	while ((l = read(toppy_fd, &ud, sizeof(ud))) == sizeof(ud)) {
		switch (ud.ud_type) {
		default:
			ft = '?';
			break;

		case UTOPPY_DIRENT_DIRECTORY:
			ft = 'd';
			break;

		case UTOPPY_DIRENT_FILE:
			ft = '-';
			break;
		}

		if ((ext = strrchr(ud.ud_path, '.')) != NULL &&
		    strcasecmp(ext, ".tap") == 0)
			ex = 'x';
		else
			ex = '-';

		tm = localtime(&ud.ud_mtime);
		strftime(tmbuf, sizeof(tmbuf), "%b %e %G %R", tm);

		printf("%crw%c %11lld %s %s\n", ft, ex, (long long)ud.ud_size,
		    tmbuf, ud.ud_path);
	}

	if (l < 0)
		err(EX_OSERR, "read(utoppy_dirent)");
}

static void
cmd_rm(int argc, char **argv)
{
	char *path;

	if (argc != 2)
		errx(EX_USAGE, "usage: rm <toppy-pathname>");

	path = argv[1];

	if (ioctl(toppy_fd, UTOPPYIODELETE, &path) < 0)
		err(EX_OSERR, "ioctl(UTOPPYIODELETE, %s)", path);
}

static void
cmd_mkdir(int argc, char **argv)
{
	char *path;

	if (argc != 2)
		errx(EX_USAGE, "usage: mkdir <toppy-pathname>");

	path = argv[1];

	if (find_toppy_dirent(path, NULL))
		errx(EX_DATAERR, "'%s' already exists", path);

	if (ioctl(toppy_fd, UTOPPYIOMKDIR, &path) < 0)
		err(EX_OSERR, "ioctl(UTOPPYIOMKDIR, %s)", path);
}

static void
cmd_rename(int argc, char **argv)
{
	struct utoppy_dirent ud;
	struct utoppy_rename ur;
	char *oldpath, *newpath, *o, *n;

	if (argc != 3)
		errx(EX_USAGE, "usage: rename <from> <to>");

	o = oldpath = argv[1];
	n = newpath = argv[2];

	for (o = oldpath; *o != '\0'; o++)
		if (*o == '\\')
			*o = '/';
	for (n = newpath; *n != '\0'; n++)
		if (*n == '\\')
			*n = '/';

	for (o = oldpath; *o && *o == '\\'; o++)
		;
	for (n = newpath; *n && *n == '\\'; n++)
		;

	if (strcmp(n, o) == 0)
		errx(EX_DATAERR, "'%s' and '%s' refer to the same file",
		    oldpath, newpath);

	if (find_toppy_dirent(oldpath, &ud) == 0)
		errx(EX_DATAERR, "'%s' does not exist on the Toppy", oldpath);

	if (ud.ud_type != UTOPPY_DIRENT_FILE)
		errx(EX_DATAERR, "%s: not a regular file", oldpath);

	if (find_toppy_dirent(newpath, &ud))
		errx(EX_DATAERR, "'%s' already exists", newpath);

	ur.ur_old_path = o;
	ur.ur_new_path = n;

	if (ioctl(toppy_fd, UTOPPYIORENAME, &ur) < 0)
		err(EX_OSERR, "ioctl(UTOPPYIORENAME, %s, %s)", oldpath,
		    newpath);
}


static void
init_progress(FILE *to, char *f, off_t fsize, off_t restart)
{
	struct ttysize ts;

	if (ioctl(fileno(to), TIOCGSIZE, &ts) == -1)
		ttywidth = 80;
	else
		ttywidth = ts.ts_cols;

	ttyout = to;
	progress = 1;
	bytes = 0;
	filesize = fsize;
	restart_point = restart;
	prefix = f;
}

static void
cmd_get(int argc, char **argv)
{
	struct utoppy_readfile ur;
	struct utoppy_dirent ud;
	struct stat st;
	char *dst, dstbuf[FILENAME_MAX];
	uint8_t *buf;
	ssize_t l;
	size_t rv;
	int ch, turbo_mode = 0, reget = 0, progbar = 0;
	FILE *ofp, *to;

	optind = 1;
	optreset = 1;

	while ((ch = getopt(argc, argv, "prt")) != -1) {
		switch (ch) {
		case 'p':
			progbar = 1;
			break;
		case 'r':
			reget = 1;
			break;
		case 't':
			turbo_mode = 1;
			break;
		default:
 get_usage:
			errx(EX_USAGE, "usage: get [-prt] <toppy-pathname> "
			    "[file | directory]");
		}
	}
	argc -= optind;
	argv += optind;

	if (argc == 1)
		dst = basename(argv[0]);
	else
	if (argc == 2) {
		dst = argv[1];
		if (stat(dst, &st) == 0 && S_ISDIR(st.st_mode)) {
			snprintf(dstbuf, sizeof(dstbuf), "%s/%s", dst,
			    basename(argv[0]));
			dst = dstbuf;
		}
	} else
		goto get_usage;

	ur.ur_path = argv[0];
	ur.ur_offset = 0;

	if ((buf = malloc(TOPPY_IO_SIZE)) == NULL)
		err(EX_OSERR, "malloc(TOPPY_IO_SIZE)");

	if (strcmp(dst, "-") == 0) {
		ofp = stdout;
		to = stderr;
		if (reget)
			warnx("Ignoring -r option in combination with stdout");
	} else {
		to = stdout;

		if (reget) {
			if (stat(dst, &st) < 0) {
				if (errno != ENOENT)
					err(EX_OSERR, "stat(%s)", dst);
			} else
			if (!S_ISREG(st.st_mode))
				errx(EX_DATAERR, "-r only works with regular "
				    "files");
			else
				ur.ur_offset = st.st_size;
		}

		if ((ofp = fopen(dst, reget ? "a" : "w")) == NULL)
			err(EX_OSERR, "fopen(%s)", dst);
	}

	if (progbar) {
		if (find_toppy_dirent(ur.ur_path, &ud) == 0)
			ud.ud_size = 0;
		init_progress(to, dst, ud.ud_size, ur.ur_offset);
	}

	if (ioctl(toppy_fd, UTOPPYIOTURBO, &turbo_mode) < 0)
		err(EX_OSERR, "ioctl(UTOPPYIOTURBO, %d)", turbo_mode);

	if (ioctl(toppy_fd, UTOPPYIOREADFILE, &ur) < 0)
		err(EX_OSERR, "ioctl(UTOPPYIOREADFILE, %s)", ur.ur_path);

	if (progbar)
		progressmeter(-1);

	for (;;) {
		while ((l = read(toppy_fd, buf, TOPPY_IO_SIZE)) < 0 &&
		    errno == EINTR)
			;

		if (l <= 0)
			break;

		rv = fwrite(buf, 1, l, ofp);

		if (rv != (size_t)l) {
			if (ofp != stdout)
				fclose(ofp);
			progressmeter(1);
			err(EX_OSERR, "fwrite(%s)", dst);
		}
		bytes += l;
	}

	if (progbar)
		progressmeter(1);

	if (ofp != stdout)
		fclose(ofp);

	if (l < 0)
		err(EX_OSERR, "read(TOPPY: ur.ur_path)");

	free(buf);
}

static void
cmd_put(int argc, char **argv)
{
	struct utoppy_writefile uw;
	struct utoppy_dirent ud;
	struct stat st;
	char dstbuf[FILENAME_MAX];
	char *src;
	void *buf;
	ssize_t rv;
	size_t l;
	int ch, turbo_mode = 0, reput = 0, progbar = 0;
	FILE *ifp;

	optind = 1;
	optreset = 1;

	while ((ch = getopt(argc, argv, "prt")) != -1) {
		switch (ch) {
		case 'p':
			progbar = 1;
			break;
		case 'r':
			reput = 1;
			break;
		case 't':
			turbo_mode = 1;
			break;
		default:
 put_usage:
			errx(EX_USAGE, "usage: put [-prt] <local-pathname> "
			    "<toppy-pathname>");
		}
	}
	argc -= optind;
	argv += optind;

	if (argc != 2)
		goto put_usage;

	src = argv[0];
	uw.uw_path = argv[1];

	if (stat(src, &st) < 0)
		err(EX_OSERR, "%s", src);

	if (!S_ISREG(st.st_mode))
		errx(EX_DATAERR, "'%s' is not a regular file", src);

	uw.uw_size = st.st_size;
	uw.uw_mtime = st.st_mtime;
	uw.uw_offset = 0;

	if (find_toppy_dirent(uw.uw_path, &ud)) {
		if (ud.ud_type == UTOPPY_DIRENT_DIRECTORY) {
			snprintf(dstbuf, sizeof(dstbuf), "%s/%s", uw.uw_path,
			    basename(src));
			uw.uw_path = dstbuf;
		} else
		if (ud.ud_type != UTOPPY_DIRENT_FILE)
			errx(EX_DATAERR, "'%s' is not a regular file.",
			    uw.uw_path);
		else
		if (reput) {
			if (ud.ud_size > uw.uw_size)
				errx(EX_DATAERR, "'%s' is already larger than "
				    "'%s'", uw.uw_path, src);

			uw.uw_size -= ud.ud_size;
			uw.uw_offset = ud.ud_size;
		}
	}

	if ((buf = malloc(TOPPY_IO_SIZE)) == NULL)
		err(EX_OSERR, "malloc(TOPPY_IO_SIZE)");

	if ((ifp = fopen(src, "r")) == NULL)
		err(EX_OSERR, "fopen(%s)", src);

	if (ioctl(toppy_fd, UTOPPYIOTURBO, &turbo_mode) < 0)
		err(EX_OSERR, "ioctl(UTOPPYIOTURBO, %d)", turbo_mode);

	if (ioctl(toppy_fd, UTOPPYIOWRITEFILE, &uw) < 0)
		err(EX_OSERR, "ioctl(UTOPPYIOWRITEFILE, %s)", uw.uw_path);

	if (progbar)
		init_progress(stdout, src, st.st_size, uw.uw_offset);

	if (progbar)
		progressmeter(-1);

	while ((l = fread(buf, 1, TOPPY_IO_SIZE, ifp)) > 0) {
		rv = write(toppy_fd, buf, l);
		if ((size_t)rv != l) {
			fclose(ifp);
			if (progbar)
				progressmeter(1);
			err(EX_OSERR, "write(TOPPY: %s)", uw.uw_path);
		}
		bytes += l;
	}

	if (progbar)
		progressmeter(1);

	if (ferror(ifp))
		err(EX_OSERR, "fread(%s)", src);

	fclose(ifp);
	free(buf);
}
